### PR TITLE
Add bulk publish and process

### PR DESCRIPTION
**Pull request by angusmcleod**
@pmusaraj This is the PR with both the bulk publish (outgoing) and bulk process (incoming) tasks. The names have changed for greater consistency

#### Publish (outoing)
```
activity_pub:publish[actor_id_or_handle,log_type]
```
#### Process (incoming)
```
activity_pub:process[actor_id_or_handle,target_actor_id_or_handle,log_type]
```
#### How I've been testing
In addition to the specs, the way I've been testing this locally is by running two separate servers and doing the following:

1. Populating a category on discourse one.

    ```
    thor populate:posts -p 10 -n 30
    ```
3. Enabling AP on the populated category on discourse one.
4. Publishing the populated category on discourse one.
5. Processing the published category on discourse one on a category on discourse two.

This is working for me on my local with a variety of topic and post counts.

There's a few more tweaks I want to make, particularly pagination on the outbox collection, so this is still in draft however it's ready for testing.

### CODE REVIEW

**Comment by angusmcleod** [comment]
@pmusaraj I've added pagination processing support and moved this out of draft (i.e. it's ready).

**Comment by pmusaraj** [comment]
Sorry for the delay here @angusmcleod, I am now testing this locally and on a test production instance.

File: lib/discourse_activity_pub/bulk/process.rb [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+            else
+              @created_topics_count += 1
+            end
+          end
+        end
+
+        if post.present?
+          result.posts_by_object_ap_id[object.stored.ap_id] = post
+
+          if actor.model.activity_pub_full_topic
```
Comment (pmusaraj) [inline lib/discourse_activity_pub/bulk/process.rb:347]: Does this mean that source and target models need to match? It looks like it from my testing so far, if I switch the setting on a target category to "Full topic", running the bulk process raises an error on line 48 just below. 

(More of a general comment for a follow up.)

**Review by pmusaraj — COMMENTED** [review]
Overall this looks great, @angusmcleod. I still have issues testing this properly locally but it has nothing to do with the PR, rather, I've discovered I think issues with proxying to the local environment. I will dig into that separately. 

Tested the info and process rake tasks in a production environment, and it all worked well. 

My only comment here is general, this adds a lot of new code for `bulk/process` and `bulk/publish` and I am wondering whether it's worth centralizing some of the logic or use the existing validations. For example, there are checks here for `activity_pub_full_topic`, could we instead rely on the model callbacks so that there is only one source of truth for those checks? 

This isn't necessarily a blocker for this PR, we could handle this in a follow-up, just wondering here if this is a doable task.

**Comment by angusmcleod** [comment]
> My only comment here is general, this adds a lot of new code for bulk/process and bulk/publish and I am wondering whether it's worth centralizing some of the logic or use the existing validations. For example, there are checks here for activity_pub_full_topic, could we instead rely on the model callbacks so that there is only one source of truth for those checks?

This is the approach I initially took, with that first PR I closed. The reason I moved to entirely separate libraries here is that you run into a few issues when trying to run the bulk process through the standard process:

1. Ordering. The standard process doesn't have to worry about where the current activity being processed is meant to sit relative to existing activities. In the bulk process this is quite tricky for a few reasons, including that you'll be processing non-topic structures. In order to create "topics" out of Notes from an ActivityPub service that doesn't have a concept of a topic (e.g. Mastodon) you have to rely on the reply chain (i.e. `inReplyTo`). The only order you receive the data in (from a services' outbox) is creation order (hopefully, not 100% reliable). Reply chain order is not the same as creation order. This means you have to process, order and then create the activities, objects and their associated Discourse models differently from the standard process.

2. Performance. The performance of running large numbers of objects through the standard process is predictably poor, as it's not really designed for that. A single Activity can result in the creation of up to 7 new objects in the database. If you have 30 posts to process that around 200 new objects in the database. If you have 10 topics of 30 posts, which is quite possible if you're importing the outbox of an existing Discourse category, that's 2000 new objects in the database. Running that, single-threaded as it were, through the standard process takes quite a while.

Given the above two issues, and a few other smaller ones, moving to a dedicated bulk library seemed to make more sense. I did then consider whether the "standard" process should not also run through the process set up for bulk processing, however I don't think that makes sense either. The standard process is inherently event-driven, and needs to be, in order to properly and stably integrate into `discourse/discourse`. This also means that the different parts of the process, e.g. the initial processing and validation of activities, then the conversion of activities into discourse objects, or, conversely, the creation of activities being triggered once discourse objects are created, need to be separated. They can't be combined in the way they need to be for the bulk process (in particular for the purpose of ordering).

File: lib/discourse_activity_pub/bulk/process.rb (hunk 1 above)
Comment (angusmcleod) [reply lib/discourse_activity_pub/bulk/process.rb:347]: I'm going to investigate this and get back to you.

Comment (angusmcleod) [reply lib/discourse_activity_pub/bulk/process.rb:349]: @pmusaraj I've added some fixes and tests for this scenario https://github.com/discourse/discourse-activity-pub/pull/72/commits/1173eaa35f671b4140f73f0bcabf2bc8d8fbc0bc

**Review by pmusaraj — APPROVED** [review]
Will merge this tomorrow am as well, thanks!

**Pull request merged**